core/templates/domain/exploration/RuleInputTypeFactory.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory to determine correct type of RuleInputs
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { IRuleInput } from 'domain/exploration/RuleObjectFactory';
+import { INumberWithUnitsBackendDict } from
+  'domain/objects/NumberWithUnitsObjectFactory';
+import { IFractionDict } from
+  'domain/objects/FractionObjectFactory.ts';
+/* eslint-disable max-len */
+import { INote } from 'extensions/interactions/MusicNotesInput/directives/music-notes-input-rules.service';
+import { IGraphBackendDict } from 'extensions/interactions/GraphInput/directives/graph-detail.service';
+/* eslint-enable max-len */
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RuleInputTypeFactory {
+  _isGraph(variable: Object): variable is IGraphBackendDict {
+    if (this._isString(variable) || this._isNumber(variable)) {
+      return false;
+    }
+
+    return 'vertices' in variable;
+  }
+
+  _isNoteArray(variable: Object): variable is INote[] {
+    if (Array.isArray(variable)) {
+      if (variable.length === 0) {
+        return true;
+      }
+
+      if (typeof variable[0] === 'string' || typeof variable[0] === 'number') {
+        return false;
+      }
+
+      return 'readableNoteName' in variable[0];
+    }
+
+    return false;
+  }
```

Comment:
Now, in these functions we had two options
- simply write `return true;`
- Actually check the type of the variable

Currently I've followed the second point.
Any can be used since we are going to use this factory only for changing the type of `IRuleInput` to the actual type.

Which one do you guys prefer?

Reply:
/cc @vojtechjelinek @ankita240796

Reply:
I'm not sure what is the difference here, in one case we would be using just `variable is INote[]` and in the second there is the actual code?

Reply:
The first point means something like this
```typescript
_isNoteArray(variable: Object): variable is INote[] {
    return true;
  }
```

Reply:
I think that the second option is fine, thanks!

Reply:
Yes, let us go with option 2.

Reply:
Hmm, due to the failure of e2e tests, I think we would have to go with the first option. Because somewhere in the codebase it is happening that the `variable` should be a number afterwards but is currently `null` or currently of no properties.
 /cc @ankita240796 @vojtechjelinek

Reply:
Also, I was thinking we could have a general `TypeChangeService` if we agree on the above point.

Reply:
> Hmm, due to the failure of e2e tests, I think we would have to go with the first option. Because somewhere in the codebase it is happening that the variable should be of some type but is currently null or at that time has no properties.

Reply:
No I think it doesn't needs to be fixed. It's something like suppose `inputs` is of type `IRuleInputs` and then `inputs.x is of type `IRuleInput` and is currently empty. And we are doing something like
```js
inputs.x.vertices = [2,2];
```
It will throw a ts error that `vertices` don't exist on type `IRuleInput`.

Now if we use our `type factory` it will throw an error as well because vertices don't exist on `inputs.x`. 

But with the first method, we can just assume that it is of type `IGraphDict`.

Reply:
Ok, thanks for explaining.